File: 600-y.rb
U = ->(f) { f.(f) }
fact = U.(
  ->(f) {
    recursion = ->(n) {
      f.(f).(n)
    }
    ->(x) {
      if x == 1
        1
      else
        x * recursion.(x-1)
      end
    }
  }
)

puts fact.(8)

#Refactor - create a new function and pass the internal function into the block as a new argument

File: js/600-y.js
const U = (f) => { return f(f) }
const fact = U((f) => {
  const recur = (n) => { return f(f)(n) }
  return (x) => {
    if (x === 0) {
      return 1
    } else {
      return x * recur(x-1)
    }
  }
})
console.log(fact(8))

//Refactor - create a new function and pass the internal function into the block as a new argument
